File: sdk-rust/src/migrator.rs
//! Tiny migration runner. Reads V###__name.sql files embedded at compile
//! time, tracks applied versions in `gatedhouse.schema_versions`, applies
//! pending migrations under a Postgres advisory lock so concurrent
//! instances don't race.

use std::collections::HashSet;

use postgres::Client;

use crate::database::Database;
use crate::error::GatedhouseError;

// Stable but arbitrary; matches the Java side so the lock is shared if
// both languages are pointed at the same database.
const ADVISORY_LOCK_KEY: i64 = 0x6761746564686F75; // 'gatedhou'

const MIGRATIONS_INDEX: &str = include_str!("migrations/migrations.txt");
const V001_SQL: &str = include_str!("migrations/V001__init.sql");

struct Migration {
    version: i32,
    name: &'static str,
    sql: &'static str,
    checksum: String,
}

pub fn migrate(database: &dyn Database) -> Result<(), GatedhouseError> {
    let migrations = available_migrations()?;

    let mut conn = database.connection()?;
    acquire_lock(&mut conn)?;
    let lock_result = (|| -> Result<(), GatedhouseError> {
        ensure_bookkeeping(&mut conn)?;
        let applied = applied_versions(&mut conn)?;
        for migration in migrations {
            if applied.contains(&migration.version) {
                continue;
            }
            apply(&mut conn, &migration)?;
        }
        Ok(())
    })();
    let release_result = release_lock(&mut conn);
    lock_result?;
    release_result?;
    Ok(())
}

fn available_migrations() -> Result<Vec<Migration>, GatedhouseError> {
    let mut out: Vec<Migration> = Vec::new();
    for raw in MIGRATIONS_INDEX.lines() {
        let line = raw.trim();
        if line.is_empty() || line.starts_with('#') {
            continue;
        }
        let (version, name, sql) = match line {
            "V001__init.sql" => (1, "init", V001_SQL),
            other => {
                return Err(GatedhouseError::Initialization(format!(
                    "Unknown migration filename in index: {other:?}. \
                     Add it as an `include_str!` constant in src/migrator.rs."
                )));
            }
        };
        let checksum = fnv1a_hex(sql);
        out.push(Migration {
            version,
            name,
            sql,
            checksum,
        });
    }
    out.sort_by_key(|m| m.version);
    Ok(out)
}

fn acquire_lock(conn: &mut Client) -> Result<(), GatedhouseError> {
    conn.execute("SELECT pg_advisory_lock($1)", &[&ADVISORY_LOCK_KEY])?;
    Ok(())
}

fn release_lock(conn: &mut Client) -> Result<(), GatedhouseError> {
    conn.execute("SELECT pg_advisory_unlock($1)", &[&ADVISORY_LOCK_KEY])?;
    Ok(())
}

fn ensure_bookkeeping(conn: &mut Client) -> Result<(), GatedhouseError> {
    conn.batch_execute(
        "CREATE SCHEMA IF NOT EXISTS gatedhouse; \
         CREATE TABLE IF NOT EXISTS gatedhouse.schema_versions ( \
             version    INTEGER PRIMARY KEY, \
             name       TEXT NOT NULL, \
             checksum   TEXT NOT NULL, \
             applied_at TIMESTAMPTZ NOT NULL DEFAULT NOW() \
         );",
    )?;
    Ok(())
}

fn applied_versions(conn: &mut Client) -> Result<HashSet<i32>, GatedhouseError> {
    let rows = conn.query("SELECT version FROM gatedhouse.schema_versions", &[])?;
    Ok(rows.into_iter().map(|r| r.get::<_, i32>(0)).collect())
}

fn apply(conn: &mut Client, migration: &Migration) -> Result<(), GatedhouseError> {
    let mut tx = conn.transaction()?;
    tx.batch_execute(migration.sql)?;
    tx.execute(
        "INSERT INTO gatedhouse.schema_versions (version, name, checksum) \
         VALUES ($1, $2, $3)",
        &[&migration.version, &migration.name, &migration.checksum],
    )?;
    tx.commit()?;
    Ok(())
}

// Tiny hand-rolled SHA-256 hex via the `sha2` crate would add another
// dep; instead, lean on jsonwebtoken's transitive `ring` to avoid
// adding anything. ring exposes `digest::SHA256`. But that's a
// fragile assumption — easier and dep-free to use a minimal hash.
//
// For the migration runner the checksum is recorded for human / future
// diagnostic use, not enforced cryptographically against tampering.
// A non-cryptographic FxHash would be enough; we use a simple hex of
// FNV-1a 64-bit to avoid any extra dep.
fn fnv1a_hex(s: &str) -> String {
    let mut h: u64 = 0xcbf29ce484222325;
    for b in s.as_bytes() {
        h ^= *b as u64;
        h = h.wrapping_mul(0x100000001b3);
    }
    format!("fnv1a64:{h:016x}")
}

File: sdk-python/gatedhouse/_migrator.py
# Copyright (c) 2026 12vectors.com
# SPDX-License-Identifier: MIT
# See the LICENSE file in the repository root for the full license text.

"""Tiny migration runner. Mirrors the Java ``Migrator`` package-private
class. Reads V###__name.sql files from the migrations resource directory,
tracks applied versions in ``gatedhouse.schema_versions``, applies
pending migrations under a Postgres advisory lock so concurrent
instances don't race.
"""

from __future__ import annotations

import hashlib
import re
from importlib.resources import files

from psycopg import errors as pg_errors

from ._database import Database
from ._exceptions import GatedhouseInitializationError

# Stable but arbitrary; matches the Java side so the lock is shared.
_ADVISORY_LOCK_KEY = 0x6761746564686F75  # 'gatedhou'

_FILENAME_PATTERN = re.compile(r"^V(\d+)__([A-Za-z0-9_]+)\.sql$")
_MIGRATIONS_PACKAGE = "gatedhouse.migrations"


def migrate(database: Database) -> None:
    available = _load_available_migrations()

    try:
        with database.connection() as conn:
            conn.autocommit = True
            with conn.cursor() as cur:
                _acquire_lock(cur)
            try:
                with conn.cursor() as cur:
                    _ensure_bookkeeping(cur)
                    applied = _applied_versions(cur)
                for migration in available:
                    if migration.version in applied:
                        continue
                    _apply(conn, migration)
            finally:
                with conn.cursor() as cur:
                    _release_lock(cur)
    except pg_errors.Error as e:
        raise GatedhouseInitializationError(
            f"Gatedhouse migration failed: {e}"
        ) from e


# ---- migration discovery -------------------------------------------------


class _Migration:

    __slots__ = ("version", "name", "sql", "checksum")

    def __init__(self, version: int, name: str, sql: str, checksum: str) -> None:
        self.version = version
        self.name = name
        self.sql = sql
        self.checksum = checksum


def _load_available_migrations() -> list[_Migration]:
    migrations_dir = files(_MIGRATIONS_PACKAGE)
    index_text = (migrations_dir / "migrations.txt").read_text(encoding="utf-8")

    out: list[_Migration] = []
    for raw in index_text.splitlines():
        line = raw.strip()
        if not line or line.startswith("#"):
            continue
        match = _FILENAME_PATTERN.match(line)
        if not match:
            raise GatedhouseInitializationError(
                f"Migration filename does not match V###__name.sql: {line!r}"
            )
        version = int(match.group(1))
        name = match.group(2)
        sql = (migrations_dir / line).read_text(encoding="utf-8")
        checksum = hashlib.sha256(sql.encode("utf-8")).hexdigest()
        out.append(_Migration(version, name, sql, checksum))
    out.sort(key=lambda m: m.version)
    return out


# ---- bookkeeping ---------------------------------------------------------


def _acquire_lock(cur) -> None:
    cur.execute("SELECT pg_advisory_lock(%s)", (_ADVISORY_LOCK_KEY,))
    cur.fetchone()


def _release_lock(cur) -> None:
    cur.execute("SELECT pg_advisory_unlock(%s)", (_ADVISORY_LOCK_KEY,))
    cur.fetchone()


def _ensure_bookkeeping(cur) -> None:
    cur.execute("CREATE SCHEMA IF NOT EXISTS gatedhouse")
    cur.execute(
        "CREATE TABLE IF NOT EXISTS gatedhouse.schema_versions ("
        "    version    INTEGER PRIMARY KEY,"
        "    name       TEXT NOT NULL,"
        "    checksum   TEXT NOT NULL,"
        "    applied_at TIMESTAMPTZ NOT NULL DEFAULT NOW()"
        ")"
    )


def _applied_versions(cur) -> set[int]:
    cur.execute("SELECT version FROM gatedhouse.schema_versions")
    return {row[0] for row in cur.fetchall()}


def _apply(conn, migration: _Migration) -> None:
    prev_autocommit = conn.autocommit
    conn.autocommit = False
    try:
        with conn.cursor() as cur:
            cur.execute(migration.sql)
            cur.execute(
                "INSERT INTO gatedhouse.schema_versions "
                "(version, name, checksum) VALUES (%s, %s, %s)",
                (migration.version, migration.name, migration.checksum),
            )
        conn.commit()
    except pg_errors.Error:
        conn.rollback()
        raise
    finally:
        conn.autocommit = prev_autocommit
